app.py

```diff
@@ -1007,6 +1007,170 @@ def set_routing_time_maximum(self):
         return success
 
 
+
+# justin
+    def matrix_reduction(self, matrix):
+        """
+        Performs the matrix reduction for branch-and-bound
+        Returns a reduced matrix
+        """
+        reduction_cost = 0
+        # Finds the minimum value to make a row have a zero
+        for key in matrix.keys():

```

Comment:
Python is pass-by-reference. So when you do:
```
for k, v in matrix.keys():
    v["value"] = 10
```
It changes the original dictionary that's passed in. But we shouldn't be doing that.

Create a copy of the matrix and set values to that:
```
temporary_matrix = matrix.copy()

for key in matrix.keys():
    ...
    for k,v in matrix.items():
        # Set the values in the temporary dictionary
        temporary_matrix[v][direc]['cost] = ...
...

# Return the matrix with the new values
return reduced_cost, temporary_matrix
```